### PR TITLE
62 feature first prototype efficientps

### PR DESCRIPTION
**Pull request by MarcoRiedenauer**
# Description

This PR includes the first working version of EfficientPS on our generated dataset. 

## Time invested

20h Marco
14h Korbinian + 16h Korbinian

## Type of change

- New feature (non-breaking change which adds functionality)

## Does this PR introduce a breaking change?

No, since nothing depends on it yet.

## Most important changes

Most important files are labels, train_net, createPanopticImgs and panoptic dataset.
dataset_converter is also important, but will be changed in #183.
Don't review visualize.py, since I just copied it in the project and didn't have a look at it yet.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works (might be obsolete with CI later on)
- [x] New and existing unit tests pass locally with my changes (might be obsolete with CI later on)

### CODE REVIEW

**Comment by nylser** [comment]
Unfortunately, there still are some problems in the evaluation due to our changed format

File: code/perception/src/panoptic_segmentation/efficientps/semantic_head/semantic_head.py [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+        self.mc_16_to_8 = MC()
+        self.mc_8_to_4 = MC()
+
+        self.last_conv = nn.Conv2d(512, nb_class, 1)
+
+        self.softmax = nn.Softmax(dim=1)
+        self.cross_entropy_loss = nn.CrossEntropyLoss(reduction='none')
+
+    def forward(self, inputs, targets={}):
+        # TODO Make a loop
```
Comment (JoKircher) [inline code/perception/src/panoptic_segmentation/efficientps/semantic_head/semantic_head.py:41]: Here is still a todo. Do you intent to make a loop?

File: code/perception/src/panoptic_segmentation/efficientps/semantic_head/semantic_head.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+        efficientps.utils.depthwise_separable_conv import \
+        DepthwiseSeparableConv
+except ImportError:
+    from efficientps.utils.depthwise_separable_conv import \
+        DepthwiseSeparableConv
+
+
+class SemanticHead(nn.Module):
+    """
+    Semantic Head compose of three main module DPC, LSFE and MC
```
Comment (JoKircher) [inline code/perception/src/panoptic_segmentation/efficientps/semantic_head/semantic_head.py:18]: What does DPC, LSFE and MC stand for?

File: code/perception/src/panoptic_segmentation/efficientps/utils/depthwise_separable_conv.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
+import torch.nn as nn
+from inplace_abn import InPlaceABN
+
+
+class DepthwiseSeparableConv(nn.Module):
+    """
+    DepthwiseSeparableConv from MobileNet, code largely inspire from mmcv
+    DepthwiseSeparableConvModule but simplify
+    TODO Improve the normalisation choices
```
Comment (JoKircher) [inline code/perception/src/panoptic_segmentation/efficientps/utils/depthwise_separable_conv.py:9]: Another todo

File: code/perception/src/segmentation_node.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+            topic=f"/carla/{self.role_name}/{self.side}/image",
+            qos_profile=1
+        )
+
+    def setup_camera_publishers(self):
+        self.publisher = self.new_publisher(
+            msg_type=numpy_msg(Image),
+            topic=f"/paf/{self.role_name}/{self.side}/segmented_image",
+            qos_profile=1
+        )
```
Comment (JoKircher) [inline code/perception/src/segmentation_node.py:65]: As far as I understood, the segmentation node only publishes the segmentated images, right? Where is the detection of traffic lights, traffic signs, cars driving in front of us, etc. handled?

**Review by JoKircher — CHANGES_REQUESTED** [review]
Thank you this PR! It is adding some much appreciated functionility. However, due to my just basic understanding of the topic there are not many in depth question I can propose. The coding style and documentation look very clean. I will approve this PR after testing. 

How can I test this PR? Just run this branch and check the published topic?

**Comment by MarcoRiedenauer** [comment]
> Thank you this PR! It is adding some much appreciated functionility. However, due to my just basic understanding of the topic there are not many in depth question I can propose. The coding style and documentation look very clean. I will approve this PR after testing.
> 
> How can I test this PR? Just run this branch and check the published topic?

Yes, should be enough for general testing. If you want to test this PR completely, you can also create a dataset with the dataset_generator and dataset_converter, convert the images with createPanopticImgs.py and train the model. But that will be too much, I think ^^

File: code/perception/src/segmentation_node.py (hunk 4 above)
Comment (MarcoRiedenauer) [reply code/perception/src/segmentation_node.py:65]: These features aren't implemented yet and not part of this PR. I think, I will create another issue for traffic light snipping and classification.

File: code/perception/src/panoptic_segmentation/efficientps/semantic_head/semantic_head.py (hunk 2 above)
Comment (MarcoRiedenauer) [reply code/perception/src/panoptic_segmentation/efficientps/semantic_head/semantic_head.py:18]: Good question ^^ We copied this code from https://github.com/vincrichard/EfficientPS, so the shortcuts and Todos are not from us. And it is not important for our task :)

File: code/perception/src/panoptic_segmentation/efficientps/semantic_head/semantic_head.py (hunk 1 above)
Comment (MarcoRiedenauer) [reply code/perception/src/panoptic_segmentation/efficientps/semantic_head/semantic_head.py]: Nope, was just a todo added by the creator of the code. Todo is removed now :)

File: code/perception/src/panoptic_segmentation/efficientps/utils/depthwise_separable_conv.py (hunk 3 above)
Comment (MarcoRiedenauer) [reply code/perception/src/panoptic_segmentation/efficientps/utils/depthwise_separable_conv.py]: Removed

**Review by JoKircher — APPROVED** [review]
Looks good to me now!
